File: apps/api/src/washflow_api/schemas/status.py
"""Status schemas."""

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class StatusCreate(BaseModel):
    """Status creation schema."""

    name: str = Field(min_length=2, max_length=40)
    is_terminal: bool = Field(default=False)

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str) -> str:
        return v.strip()


class StatusUpdate(BaseModel):
    """Status update schema."""

    name: str | None = Field(default=None, min_length=2, max_length=40)
    is_terminal: bool | None = None

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str | None) -> str | None:
        if v is not None:
            return v.strip()
        return v


class StatusOut(BaseModel):
    """Status response schema."""

    id: str
    organization_id: str
    location_id: str
    name: str
    sort_order: int
    is_terminal: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class StatusListResponse(BaseModel):
    """Paginated status list response."""

    items: list[StatusOut]
    total: int


class StatusReorderRequest(BaseModel):
    """Request to reorder statuses."""

    ordered_ids: list[str] = Field(min_length=1)

File: apps/web/src/stores/status.ts
import { defineStore } from 'pinia'
import { ref, watch } from 'vue'
import api from '@/lib/api'
import { useLocationStore } from './location'

export interface Status {
  id: string
  organization_id: string
  location_id: string
  name: string
  sort_order: number
  is_terminal: boolean
  created_at: string
  updated_at: string
}

export interface StatusListResponse {
  items: Status[]
  total: number
}

export interface StatusCreate {
  name: string
  is_terminal?: boolean
}

export interface StatusUpdate {
  name?: string
  is_terminal?: boolean
}

export const useStatusStore = defineStore('status', () => {
  const locationStore = useLocationStore()

  const statuses = ref<Status[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchStatuses() {
    if (!locationStore.selectedLocationId) {
      statuses.value = []
      return
    }

    loading.value = true
    error.value = null
    try {
      const response = await api.get<StatusListResponse>(
        `/api/v1/locations/${locationStore.selectedLocationId}/statuses`
      )
      statuses.value = response.data.items
    } catch (err) {
      error.value = 'Error al cargar estados'
      console.error('Failed to fetch statuses:', err)
    } finally {
      loading.value = false
    }
  }

  async function createStatus(data: StatusCreate) {
    if (!locationStore.selectedLocationId) {
      throw new Error('No hay sucursal seleccionada')
    }

    const response = await api.post<Status>(
      `/api/v1/locations/${locationStore.selectedLocationId}/statuses`,
      data
    )
    statuses.value.push(response.data)
    // Re-sort after adding
    statuses.value.sort((a, b) => a.sort_order - b.sort_order)
    return response.data
  }

  async function updateStatus(id: string, data: StatusUpdate) {
    const response = await api.patch<Status>(`/api/v1/statuses/${id}`, data)
    const index = statuses.value.findIndex((s) => s.id === id)
    if (index !== -1) {
      statuses.value[index] = response.data
    }
    return response.data
  }

  async function reorderStatuses(orderedIds: string[]) {
    if (!locationStore.selectedLocationId) {
      throw new Error('No hay sucursal seleccionada')
    }

    const response = await api.post<StatusListResponse>(
      `/api/v1/locations/${locationStore.selectedLocationId}/statuses/reorder`,
      { ordered_ids: orderedIds }
    )
    statuses.value = response.data.items
    return response.data.items
  }

  // Refetch when location changes
  watch(
    () => locationStore.selectedLocationId,
    () => {
      fetchStatuses()
    },
    { immediate: true }
  )

  return {
    statuses,
    loading,
    error,
    fetchStatuses,
    createStatus,
    updateStatus,
    reorderStatuses,
  }
})
